main/src/ca/uwaterloo/flix/language/phase/unification/FastBoolUnification.scala

```diff
@@ -0,0 +1,1715 @@
+        val termVars = vars0.map(_.x)
+
+        // Extend trueCsts and trueVars.
+        val currentCsts = trueCsts ++ termCsts
+        val currentVars = trueVars ++ termVars
+
+        // Recurse on the sub-terms with the updated maps.
+        val rest = rest0.collect {
+          case t: Term.Not => visit(t, currentCsts, currentVars)
+          case t: Term.And => visit(t, currentCsts, currentVars)
+          case t: Term.Or => visit(t, currentCsts, currentVars)
+          case _: Term.Cst =>
+            // Cannot happen because the invariant of [[Term.mkAnd]] ensures there are no constants in `rest`.
+            throw InternalCompilerException("Unexpected constant", SourceLocation.Unknown)
+          case _: Term.Var =>
+            // Cannot happen because the invariant of [[Term.mkAnd]] ensures there are no variables in `rest`.
+            throw InternalCompilerException("Unexpected variable", SourceLocation.Unknown)
+        }
+
+        // Compute new constant and variable sets by removing constants and variables that hold.
+        val csts = termCsts -- trueCsts
+        val vars = termVars -- trueVars
+
+        // Recompose the conjunction. We use the smart constructor because some sets may have become empty.
+        Term.mkAnd(csts.toList.map(Term.Cst) ++ vars.toList.map(Term.Var) ++ rest)
+
+      case Term.Or(ts) => Term.mkOr(ts.map(visit(_, trueCsts, trueVars)))
+    }
+
+    visit(t, SortedSet.empty, SortedSet.empty)
+  }
+
+  /**
+    * Companion object for [[Equation]].
+    */
+  object Equation {
+    /**
+      * Returns a unification equation  `t1 ~ t2` between the terms `t1` and `t2`.
+      *
+      * The smart constructor performs normalization:
+      * - We move true and false to the rhs.
+      * - We move a variable to the lhs (unconditionally).
+      * - We reorder constant/variables equations so that the smaller constant/variable is on the lhs.
+      *
+      * Examples:
+      * -     true ~ x7 ==> x7 ~ true
+      * -       c3 ~ c2 ==> c2 ~ c3
+      * -       x7 ~ x5 ==> x5 ~ x7
+      * - x3 /\ x7 ~ x4 ==> x4 ~ x3 /\ x7
+      * - x1 /\ x2 ~ c5 ==> c5 ~ x1 /\ x2
+      */
+    def mk(t1: Term, t2: Term, loc: SourceLocation): Equation = (t1, t2) match {
+      case (Term.Cst(c1), Term.Cst(c2)) => if (c1 <= c2) Equation(t1, t2, loc) else Equation(t2, t1, loc)
+      case (Term.Var(x1), Term.Var(x2)) => if (x1 <= x2) Equation(t1, t2, loc) else Equation(t2, t1, loc)
+      case (Term.True, _) => Equation(t2, Term.True, loc)
+      case (Term.False, _) => Equation(t2, Term.False, loc)
+      case (_, Term.Var(_)) => Equation(t2, t1, loc)
+      case _ => Equation(t1, t2, loc)
+    }
+  }
+
+  /**
+    * Represents a unification equation `t1 ~ t2` between the terms `t1` and `t2`.
+    *
+    * WARNING: Equations should be normalized. Use the smart constructor [[Equation.mk]] to create a new equation.
+    */
+  case class Equation(t1: Term, t2: Term, loc: SourceLocation) {
```

Comment:
`case class Equation private(t1: Term, t2: Term, loc: SourceLocation) {` is stronger than a comment

Reply:
I did not know you can do that! I added private, but kept the warning.